### PR TITLE
Introduce "composer-exit-on-patch-failure" config option.

### PR DESCRIPTION
**Pull request by Leksat**
#40

### CODE REVIEW

**Comment by cweagans** [comment]
This will cause `composer validate` to fail for anyone using this option. I'd prefer not to break this behavior, as some people run `composer validate` as part of their test suite. Having tests suddenly start failing would not be a great thing.

**Comment by Leksat** [comment]
@cweagans, I can't get what you mean... For me `composer validate` works just fine with `"composer-exit-on-patch-failure": true`.

File: src/Patches.php [hunk 1]
```diff
@@ -269,10 +269,12 @@ public function postInstall(PackageEvent $event) {
         $extra['patches_applied'][$description] = $url;
       }
       catch (\Exception $e) {
-        $this->io->write('   <error>Could not apply patch! Skipping.</error>');
-        if (getenv('COMPOSER_EXIT_ON_PATCH_FAILURE')) {
```
Comment (bartfeenstra) [inline src/Patches.php]: Removing support for this constant breaks backwards compatibility.

**Comment by cweagans** [comment]
According to the [schema](https://github.com/composer/composer/blob/master/res/composer-schema.json), that should fail. I was going to open a PR against Composer to relax the schema for `config` so that plugins can use it for flags like this, but I can't get `composer validate` to fail now that you mention it. Maybe a bug in the schema validator that Composer uses? I'll investigate more later.

File: src/Patches.php (hunk 1 above)
Comment (cweagans) [inline src/Patches.php]: 👍  ^^ what @bartfeenstra said. Thanks for catching that.

**Comment by Leksat** [comment]
I also restored the `Could not apply patch! Skipping.` message for the case if someone parses the composer output.

File: src/Patches.php (hunk 1 above)
Comment (Leksat) [inline src/Patches.php]: Fixed :wink:

**Comment by danepowell** [comment]
Would the `extra` key be a better alternative to `config`? From the schema docs:

> extra: Arbitrary extra data that can be used by plugins, for example, package of type composer-plugin may have a 'class' key defining an installer class name.

**Comment by Leksat** [comment]
@cweagans, thanks for the explanation. I did not know that `config` options are limited. I moved the option to the `extra` section.

**Comment by erikhansen** [comment]
@Leksat I would love to be able to use this feature. Would you be able to resolve the merge conflict with this PR so that it can be merged in?

**Comment by dan2k3k4** [comment]
Hey @erikhansen, I updated this PR to resolve the conflict. Please test and provide feedback. Thanks in advance!

**Comment by erikhansen** [comment]
@cweagans - Based on my testing, I'd recommend merging this PR.

@dan2k3k4 - I finally made time to test your PR and it's working great on my end. Screenshots demonstrating things working:

*composer.json file*

09-47-08 composer json sites git develop -8iox7

*Exception being thrown and non-zero exit code*
09-45-29 healthyback dev -bash 176x49-loc0b

**Comment by erikhansen** [comment]
@cweagans Any update on this getting included in the next release? Can we expect it in your 2.0 update?

**Comment by cweagans** [comment]
Sorry for the delay, folks.

**Pull request merged**